### ISSUE TITLE
How you divide the train step into Orthogonality step and Gradient step?

### ISSUE THREAD

**Issue by lvyanxuan** [KlugerLab/SpectralNet#10]
Hello, I read your paper and was attracted by the method you use on SpeatralNet's training process. But I can't find the corresponding code in your project.
In my understanding, during the training process of the spectral net, you modify the Orthogonality layer's weight in one epoch, then BP other layer's weight in the next epoch. But I find that your spectral net doesn't distinguish epochs? And all the net layers was trained with the same input data?
could you please explain how the alternate training is implemented？thank you.

**Comment by lihenryhfl** [comment]
Hi, thanks for your question. The layers are trained individually on each training step, not epoch. As a result, we can just train them at the same time, but with different training samples.

SpectralNet takes in three different types of input, all of which are tensorflow placeholders stored in a dict called "inputs" (created at [1]). The regular training input goes into a layer stored with the key 'Unlabeled'. The orthogonality layer's inputs are called 'Orthonorm'. Finally, this is never currently used, 
but there is an input called 'Labeled' for semi-supervised learning (which is still in the works).

If you notice here [2], there is a loop in train_step that goes through each of the three input types, and assigns a random minibatch to each placeholder. The 'Orthonorm' placeholder, in particular, is connected to the orthogonalization machinery here [3].

And that's how the orthonorm layer is trained separately from SpectralNet for each training step, but inside the same train_step call.

Let me know if you have any questions.

[1]: https://github.com/KlugerLab/SpectralNet/blob/ce99307a9b86ec5d201bf9cd56cf34397e341814/src/applications/spectralnet.py#L61

[2]: https://github.com/KlugerLab/SpectralNet/blob/ce99307a9b86ec5d201bf9cd56cf34397e341814/src/core/train.py#L68

[3]:
https://github.com/KlugerLab/SpectralNet/blob/ce99307a9b86ec5d201bf9cd56cf34397e341814/src/core/layer.py#L104

**Comment by lvyanxuan** [comment]
Thank you！I got it.
